### PR TITLE
Update deposits processing v0.6  - Draft PR

### PR DESCRIPTION
**Pull request by cyberbono3**
This PR updates [deposits processing](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#deposits)

Changed:
1)Updated ProcessDeposit  in validator.go
2)Added tests for ProcessDeposit in validator_test.go:  
TestProcessDeposit_DepositIndexNotMatchStateDepositIndex
TestProcessDeposit_DepositPublicKeyNotInValidatorsPubKeys
TestProcessDeposit_DepositPublicKeyInValidatorsPubKeys

[Part of] #2307

---
Need help

Issues:

1.[verify_merkle_branch](https://github.com/prysmaticlabs/prysm/blob/c33575d51dc4ee1ac5fb830af16c467b950590bd/beacon-chain/core/validators/validator.go#L154) in spec  does not match [VerifyMerkleProof implementation](https://github.com/prysmaticlabs/prysm/blob/spec-v0.6/shared/trieutil/sparse_merkle.go#L49)

2. [bls_verify](https://github.com/prysmaticlabs/prysm/blob/c33575d51dc4ee1ac5fb830af16c467b950590bd/beacon-chain/core/validators/validator.go#L191) in spec does not match [bls.Verify implementation](https://github.com/prysmaticlabs/prysm/blob/spec-v0.6/shared/bls/bls.go#L98)

3. bazel test output "/beacon-chain/core/validators/validator.go:161:15: multiple-value ssz.TreeHash() in single-value context". What method should be used to compute  leaf=hash_tree_root(deposit.data) ?

4. bazel test output "/beacon-chain/core/validators/validator.go:188:6: undefined: "github.com/prysmaticlabs/prysm/shared/bls".Verify" . How to invoke bls.Verify() method properly?

### CODE REVIEW

**Comment by terencechain** [comment]
@cyberbono3 let's try merge this before we open more PRs. We don't want too many outstanding PRs

**Comment by cyberbono3** [comment]
@terencechain  it is a good idea. I would appreciate any suggestions how to resolve the issues above

**Comment by terencechain** [comment]
> @terencechain it is a good idea. I would appreciate any suggestions how to resolve the issues above

Can you check the spec for those issues? they seem self explanatory, I'm not quite sure what 2, 3 and 4 are asking

**Comment by cyberbono3** [comment]
I get this error "Hashing of deposit.Data failed" by running [depositDataBuf32, err := hashutil.HashProto(deposit.Data)](https://github.com/cyberbono3/prysm/blob/update-deposits-processing-v0.6/beacon-chain/core/validators/validator.go#L151). Does anyone know why?

**Comment by nisdas** [comment]
@cyberbono3 please look at how we processed deposits before, the only difference with the spec is that the two functions were merged into one. You check the merkle proof of the deposit twice, and check for non-nil errors which have just been initialized.

File: beacon-chain/core/blocks/block_operations.go [hunk 1]
```diff
@@ -681,11 +680,7 @@ func ProcessValidatorDeposits(
 		)
 	}
 	var err error
-	var depositInput *pb.DepositInput
-	validatorIndexMap := stateutils.ValidatorIndexMap(beaconState)
 	for idx, deposit := range deposits {
-		depositData := deposit.DepositData
-		depositInput, err = helpers.DecodeDepositInput(depositData)
 		if err != nil {
```
Comment (nisdas) [inline beacon-chain/core/blocks/block_operations.go]: it makes no sense to check if err != nil, as it just has been initalized above

File: beacon-chain/core/validators/validator.go [hunk 2]
```diff
@@ -100,50 +101,105 @@ func AttestingValidatorIndices(
...
+	if deposit.Index != state.DepositIndex {
+		return nil, fmt.Errorf("deposit.Index : %d does not match state.DepositIndex: %d", deposit.Index, state.DepositIndex)
 	}
 	state.DepositIndex++
-
+  
+	pubKey := deposit.Data.Pubkey
+	amount := deposit.Data.Amount
+	validatorsPubKeys := make([][]byte, 0, len(state.ValidatorRegistry))
+	for _, v := range state.ValidatorRegistry {
```
Comment (nisdas) [inline beacon-chain/core/validators/validator.go]: why loop through here twice ?

**Comment by cyberbono3** [comment]
[test output](https://play.golang.org/p/B_GtONqi8I7) unveils the problems with HashProto(deposit.Data)

**Comment by cyberbono3** [comment]
what field of deposit.Data is it expected [here](https://github.com/cyberbono3/prysm/blob/update-deposits-processing-v0.6/beacon-chain/core/validators/validator.go#L151) ? according to VerifyMerkleProof  method, it should be a byte slice.

**Comment by cyberbono3** [comment]
Done by #2647

**Pull request closed without merging**